### PR TITLE
Soil mechanics

### PR DESCRIPTION
**Pull request by SawneyX**
added better soil mechanics (optional global variable), gradient colors for dirt height

### CODE REVIEW

File: terra/state.py [hunk 1]
```diff
@@ -26,6 +26,10 @@
 from terra.settings import IntMap
 from terra.utils import wrap_angle_rad
 
+# Flag to enable soil mechanics
+ENABLE_SOIL_MECHANICS = True
+# If TRUE: Call _apply_dump() with use_condensed_dump=True to use a more concentrated dump that works better for soil mechanics
+
```
[the remark this answers is not in the document]
Comment (Stachq1) [reply terra/state.py]: To be honest I would even consider getting rid of this whatsoever. Do we need the old soil for anything if the new one is working fine? If not then just get rid of it :)

File: terra/state.py [hunk 2]
```diff
@@ -719,13 +724,94 @@ def _apply_dig_mask(
...
+        #Optionally apply soil mechanics using the global flag
+        def apply_soil_mech():
+            map_2d = new_flattened_map.reshape(self.world.action_map.map.shape)
+            dig_mask_2d = dig_mask.reshape(self.world.action_map.map.shape)
+            return self._apply_local_soil_mechanics_simplified(map_2d, dig_mask_2d).reshape(-1)
+        return jax.lax.cond(
+            ENABLE_SOIL_MECHANICS,
+            apply_soil_mech,
+            lambda: new_flattened_map
+        )
```
Comment (Stachq1) [inline terra/state.py]: Again, honestly just get rid of this `ENABLE_SOIL_MECHANICS` flag. No need for old soil and it simplifies the logic which is always nice

File: terra/state.py [hunk 3]
```diff
@@ -744,32 +831,99 @@ def _apply_dump_mask(
...
+            # Optionally apply soil mechanics using the global flag
+            def apply_soil_mech():
+                map_2d = simple_result.reshape(self.world.action_map.map.shape)
+                mask_2d = dump_mask_final.reshape(self.world.action_map.map.shape)
+                return self._apply_local_soil_mechanics_simplified(map_2d, self._expand_mask_for_soil_mechanics(mask_2d)).reshape(-1)
+            return jax.lax.cond(
+                ENABLE_SOIL_MECHANICS,
+                apply_soil_mech,
+                lambda: simple_result
+            )
```
Comment (Stachq1) [inline terra/state.py]: Same as above - the variable is not necessary, just use the new soil. Also could we maybe change the naming a bit? `map_2d` and then `mask_2d` does not really say what this stuff is doing, its kinda hard to read haha

File: terra/state.py [hunk 4]
```diff
@@ -744,32 +831,99 @@ def _apply_dump_mask(
...
+                concentrated_mask_flat.size + 1
+            )
+            bonus_mask_flat = bonus_indices <= remaining_concentrated_volume
+            bonus_volume_2d = bonus_mask_flat.reshape(map_2d_shape).astype(IntMap)
+            new_map_2d = flattened_map.reshape(map_2d_shape).astype(IntMap) + volume_per_tile_2d + bonus_volume_2d
+            final_map_2d = jax.lax.cond(
+                ENABLE_SOIL_MECHANICS,
+                lambda: self._apply_local_soil_mechanics_simplified(new_map_2d, self._expand_mask_for_soil_mechanics(final_concentrated_mask_2d)),
+                lambda: new_map_2d
+            )
```
Comment (Stachq1) [inline terra/state.py]: Again, just always enable them

Comment (Stachq1) [reply terra/state.py]: Also can we avoid this function noodle call here?

`self._apply_local_soil_mechanics_simplified(new_map_2d, self._expand_mask_for_soil_mechanics(final_concentrated_mask_2d))`

Can we just fuse this into one function?

File: terra/state.py [hunk 5]
```diff
@@ -719,13 +724,94 @@ def _apply_dig_mask(
...
+                    return jnp.zeros_like(action_map, dtype=map_2d.dtype)
+                # JAX-compatible shape check using jnp.where for static shape determination
+                is_valid_2d = (jnp.ndim(map_2d) == 2) & (jnp.shape(map_2d)[0] > 0) & (jnp.shape(map_2d)[1] > 0)
+                return jax.lax.cond(is_valid_2d, handle_valid_2d, handle_invalid_shape, map_2d)
+            map_2d = jax.lax.fori_loop(0, n_iters, collapse_step, map_2d)
+            return map_2d.astype(action_map.dtype)
+        has_affected = jnp.any(affected_mask)
+        def do_collapse(_):
+            return collapse_body(action_map, affected_mask)
+        return jax.lax.cond(has_affected, do_collapse, lambda _: action_map, operand=None)
```
Comment (Stachq1) [inline terra/state.py]: This function is extremely hard to read due to this function in function in function ... I know that Claude does most of the reading now but still

Would it be a big problem to somehow simplify this logic maybe?

**Review by Stachq1 — CHANGES_REQUESTED** [review]
Please remove the `ENABLE_SOIL_MECHANICS` or at least move it to config and try to make less "function-in-function-in-function" calling where possible. Otherwise LGTM, nice job!

**Comment by SawneyX** [comment]
I removed the global flag and made the functions more readable with less nested functions. Also updated some variable names

**Review by Stachq1 — APPROVED** [review]
